syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,987 @@
+
+    @with_session
+    def create_batch(self, session, user, collection):
+        """Creates a batch in batch_uploads table"""
+        userid = user_key(user)
+        collectionid = self._get_collection_id(collection)
+        # Use a more precise timestamp for the batch id (timestamp in
+        # milliseconds)
+        id = ts2bigint((session.timestamp_precise.replace(tzinfo=None)
+                        - EPOCH).total_seconds())
+        try:
+            session.transaction.execute_update(
+                """\
+                INSERT INTO batches (userid, collection, id, bsos, expiry)
+                VALUES (@userid, @collectionid, @id, @bsos, @expiry)
+                """,
+                params={"userid": userid,
+                        "collectionid": collectionid,
+                        "id": ts2dt(id / 1000.0),
+                        "bsos": "",
+                        "expiry": ts2dt(session.timestamp) +
+                        datetime.timedelta(seconds=BATCH_LIFETIME)},
+                param_types={"userid": param_types.STRING,
+                             "collectionid": param_types.INT64,
+                             "id": param_types.TIMESTAMP,
+                             "bsos": param_types.STRING,
+                             "expiry": param_types.TIMESTAMP}
+            )
+        except AlreadyExists:
+            # The user tried to create two batches with the same timestamp.
+            raise ConflictError
+        return id
+
+    @with_session
+    def valid_batch(self, session, user, collection, batchid):
+        """Checks to see if the batch ID is valid and still open"""
+        userid = user_key(user)
+        collectionid = self._get_collection_id(collection)
+        batchid = ts2dt(batchid / 1000.0)
+        q = """\
+        SELECT id FROM batches
+        WHERE userid=@userid AND collection=@collectionid AND id=@id"""
+        result = session.transaction.execute_sql(
+            q,
+            params={"userid": userid,
+                    "collectionid": collectionid,
+                    "id": batchid},
+            param_types={"userid": param_types.STRING,
+                         "collectionid": param_types.INT64,
+                         "id": param_types.TIMESTAMP}
+        ).one_or_none()
+        return bool(result)
+
+    @metrics_timer("syncstorage.storage.sql.append_items_to_batch")
+    @with_session
+    def append_items_to_batch(self, session, user, collection, batchid,
+                              items):
+        """Inserts items into batch_upload_items"""
+        userid = user_key(user)
+        collectionid = self._get_collection_id(collection)
+        batchid = ts2dt(batchid / 1000.0)
+        items_json = [json.dumps(item) for item in items]
+        bsos = "%s%s" % ("\n".join(items_json), "\n" if items_json else "")
+        result = session.transaction.execute_update(
+            """\
+            UPDATE batches SET bsos = CONCAT(bsos, @bsos)
```

Comment:
Can we detect the case where `bsos` exceeds the max size limit on a string, and error out cleanly?

Reply:
We can set a pessimistically low limit on batches to prevent it, that's about it I think.

Reply:
Adding a max_total_bytes setting of 2.2 MB to ensure there's JSON serialization space around it and it will still fit in 2.5 MB.